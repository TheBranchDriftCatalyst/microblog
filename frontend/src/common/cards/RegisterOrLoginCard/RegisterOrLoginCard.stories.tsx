import { Meta } from '@storybook/react';
// import NavigationItem, { _sampleLinkObjects } from "../NavigationHeader/NavigationItem";
import RegisterOrLoginCard from './RegisterOrLoginCard'; // Adjust the import path as needed

// Define the metadata for the story
const meta: Meta<typeof RegisterOrLoginCard> = {
  title: 'Cards/RegisterOrLoginCard',
  component: RegisterOrLoginCard,
  parameters: {
    layout: 'centered'
  }
};

export default meta;

export const Default = {
  args: {
    oidcProviders: [
      {
        name: 'GitHub',
        onClick: () => {},
        // icon: ,
      },
      {
        name: 'Google',
        onClick: () => {},
        // icon: ,
      },
    ],
  },
};
