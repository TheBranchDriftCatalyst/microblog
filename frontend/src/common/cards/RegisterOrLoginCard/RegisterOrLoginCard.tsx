"use client";

import { Button } from "@/common/ui/button";
import { Card, CardContent, CardDescription, CardFooter, CardHeader, CardTitle } from "@/common/ui/card";
import { SiGit, SiGoogle } from "@icons-pack/react-simple-icons";
import { toLower } from "lodash";
import { FileQuestionIcon } from "lucide-react";
import React from "react";

import { useForm } from "react-hook-form";
import { z } from "zod";

import {
  Form,
  // Form,
  FormControl,
  FormField,
  FormItem,
  FormLabel,
  FormMessage,
} from "@/common/ui/form";
import { Input } from "@/common/ui/input";
import { Typography } from "@/common/ui/typography";
import { zodResolver } from "@hookform/resolvers/zod";

const InputFormSchema = z.object({
  username: z.string().min(2, {
    message: "Username must be at least 2 characters.",
  }),
  password: z.string().min(6, {
    message: "Username must be at least 6 characters.",
  }),
});

const ProviderIcons = {
  github: SiGit,
  google: SiGoogle,
};

interface OIDCProviderShape {
  name: keyof typeof ProviderIcons;
  onClick: () => void;
  icon?: React.ComponentType;
}

interface RegisterOrLoginCardProps {
  oidcProviders: OIDCProviderShape[];
  onLogin: (values: z.infer<typeof InputFormSchema>) => void;
  onCreateAccount: () => void;
}

// @ts-ignore
export const RegisterOrLoginCard = ({ oidcProviders, onLogin, onCreateAccount }: RegisterOrLoginCardProps) => {
  const form = useForm<z.infer<typeof InputFormSchema>>({
    resolver: zodResolver(InputFormSchema),
    defaultValues: {
      username: "",
      password: "",
    },
  });

  return (
    <Card className="w-[350px]">
      <CardHeader>
        <CardTitle>Login</CardTitle>
        <CardDescription>Enter your email below to login</CardDescription>
      </CardHeader>
      <Form {...form}>
        <form onSubmit={form.handleSubmit(onLogin)}>
          <CardContent>
            <div className="flex justify-evenly">
              {oidcProviders.map((provider) => (
                <OIDCButton key={provider.name} provider={provider} />
              ))}
            </div>
            <div className="relative mt-4">
              <div className="absolute inset-0 flex items-center">
                <span className="w-full border-t" />
              </div>
              <div className="relative flex justify-center text-xs uppercase">
                <span className="bg-background px-2 text-muted-foreground">Or continue with</span>
              </div>
            </div>
            <FormField
              control={form.control}
              name="username"
              render={({ field }) => {
                // console.log(field);
                return (
                  <FormItem>
                    <FormLabel>Username</FormLabel>
                    <FormControl>
                      <Input placeholder="someone@somewhere.com" {...field} />
                    </FormControl>
                    <FormMessage />
                  </FormItem>
                );
              }}
            />
            <FormField
              control={form.control}
              name="password"
              render={({ field }) => {
                // console.log(field);
                return (
                  <FormItem>
                    <FormLabel>Password</FormLabel>
                    <FormControl>
                      <Input type="password" {...field} />
                    </FormControl>
                    <FormMessage />
                  </FormItem>
                );
              }}
            />
          </CardContent>
          <CardFooter>
            <div className="w-full flex justify-evenly">
              <span className="shrink-1">
                <Button variant="default" type="submit">
                  Login
                </Button>
              </span>
              <Typography size="2xs" className="text-muted-foreground">
                or
              </Typography>
              <span className="shrink-1">
                <Button variant="secondary" onClick={form.handleSubmit(onCreateAccount)}>
                  Create account
                </Button>
              </span>
            </div>
          </CardFooter>
        </form>
      </Form>
    </Card>
  );
};

const OIDCButton: React.FC<{ provider: OIDCProviderShape }> = ({ provider }) => {
  const lookupName = toLower(provider.name) as keyof typeof ProviderIcons;
  const IconComponent = ProviderIcons[lookupName] || provider?.icon || FileQuestionIcon;

  return (
    <Button variant="outline" onClick={provider.onClick}>
      {IconComponent && <IconComponent className="mr-2 h-4 w-4" />}
      {provider.name.charAt(0).toUpperCase() + provider.name.slice(1)}
    </Button>
  );
};

export default RegisterOrLoginCard;
