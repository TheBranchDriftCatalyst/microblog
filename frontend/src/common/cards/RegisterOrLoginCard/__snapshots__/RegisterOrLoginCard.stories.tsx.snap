// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cards/RegisterOrLoginCard Default smoke-test 1`] = `
<div class="rounded-lg border bg-card text-card-foreground shadow-sm w-[350px]">
  <div class="flex flex-col space-y-1.5 p-6">
    <h3 class="text-2xl font-semibold leading-none tracking-tight">
      Login
    </h3>
    <p class="text-sm text-muted-foreground">
      Enter your email below to login
    </p>
  </div>
  <form>
    <div class="p-6 pt-0">
      <div class="flex justify-evenly">
        <button class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border border-input bg-background hover:bg-accent hover:text-accent-foreground h-10 px-4 py-2">
          <svg xmlns="http://www.w3.org/2000/svg"
               width="24"
               height="24"
               fill="currentColor"
               viewbox="0 0 24 24"
               class="mr-2 h-4 w-4"
          >
            <title>
              Git
            </title>
            <path d="M23.546 10.93L13.067.452c-.604-.603-1.582-.603-2.188 0L8.708 2.627l2.76 2.76c.645-.215 1.379-.07 1.889.441.516.515.658 1.258.438 1.9l2.658 2.66c.645-.223 1.387-.078 1.9.435.721.72.721 1.884 0 2.604-.719.719-1.881.719-2.6 0-.539-.541-.674-1.337-.404-1.996L12.86 8.955v6.525c.176.086.342.203.488.348.713.721.713 1.883 0 2.6-.719.721-1.889.721-2.609 0-.719-.719-.719-1.879 0-2.598.182-.18.387-.316.605-.406V8.835c-.217-.091-.424-.222-.6-.401-.545-.545-.676-1.342-.396-2.009L7.636 3.7.45 10.881c-.6.605-.6 1.584 0 2.189l10.48 10.477c.604.604 1.582.604 2.186 0l10.43-10.43c.605-.603.605-1.582 0-2.187">
            </path>
          </svg>
          GitHub
        </button>
        <button class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border border-input bg-background hover:bg-accent hover:text-accent-foreground h-10 px-4 py-2">
          <svg xmlns="http://www.w3.org/2000/svg"
               width="24"
               height="24"
               fill="currentColor"
               viewbox="0 0 24 24"
               class="mr-2 h-4 w-4"
          >
            <title>
              Google
            </title>
            <path d="M12.48 10.92v3.28h7.84c-.24 1.84-.853 3.187-1.787 4.133-1.147 1.147-2.933 2.4-6.053 2.4-4.827 0-8.6-3.893-8.6-8.72s3.773-8.72 8.6-8.72c2.6 0 4.507 1.027 5.907 2.347l2.307-2.307C18.747 1.44 16.133 0 12.48 0 5.867 0 .307 5.387.307 12s5.56 12 12.173 12c3.573 0 6.267-1.173 8.373-3.36 2.16-2.16 2.84-5.213 2.84-7.667 0-.76-.053-1.467-.173-2.053H12.48z">
            </path>
          </svg>
          Google
        </button>
      </div>
      <div class="relative mt-4">
        <div class="absolute inset-0 flex items-center">
          <span class="w-full border-t">
          </span>
        </div>
        <div class="relative flex justify-center text-xs uppercase">
          <span class="bg-background px-2 text-muted-foreground">
            Or continue with
          </span>
        </div>
      </div>
      <div class="space-y-2">
        <label class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
               for=":r0:-form-item"
        >
          Username
        </label>
        <input class="flex h-10 w-full rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50"
               placeholder="someone@somewhere.com"
               id=":r0:-form-item"
               aria-describedby=":r0:-form-item-description"
               aria-invalid="false"
               value
               name="username"
        >
      </div>
      <div class="space-y-2">
        <label class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
               for=":r1:-form-item"
        >
          Password
        </label>
        <input class="flex h-10 w-full rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50"
               id=":r1:-form-item"
               aria-describedby=":r1:-form-item-description"
               aria-invalid="false"
               type="password"
               value
               name="password"
        >
      </div>
    </div>
    <div class="flex items-center p-6 pt-0">
      <div class="w-full flex justify-evenly">
        <span class="shrink-1">
          <button class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2"
                  type="submit"
          >
            Login
          </button>
        </span>
        <p class="text-2xs sm:text-xs md:text-sm text-muted-foreground">
          or
        </p>
        <span class="shrink-1">
          <button class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-secondary text-secondary-foreground hover:bg-secondary/80 h-10 px-4 py-2">
            Create account
          </button>
        </span>
      </div>
    </div>
  </form>
</div>
`;
